---
title: "LIFE750 Advanced Genomics assessment 1 code"
author: "James C (Group 4)"
date: "2026-2-24"
output:
  html_document: default
  pdf_document: default
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE)
```
## Introduction
This is an R Markdown document containing the code used 
to generate the plots shown in our poster for the LIFE750
Advanced Genomics semester 1 assignment. 

The code contains bash-scripts used to run the 
command-line software utilised throughout. 

## Part 1
## Environment set-up
```{bash, eval=FALSE, echo=TRUE}
#miniforge was installed
wget https://github.com/conda-forge/miniforge
bash Miniforge3-Linux-x86_64.sh
conda config --set auto_activate_base false
#a dedicated environment was set-up
mamba create -n shotgun_meta python=3.11
#specified python version to use as 3.11
mamba activate shotgun_meta
mamba install -c bioconda fastqc trim_galore multiqc bowtie2 kraken2 \
krona bracken lefse flash megahit quast metabat2 prokka bbmap circos
#the krona database was updated using:
ktUpdateTaxonomy.sh
```

## Kraken2 set-up
```bash
#in home directory
mkdir kraken2_db
cd kraken2_db
wget https://genome-idx.s3.amazonaws.com/kraken/k2_standard_08_GB_20251015.tar.gz
tar -xvzf k2_standard_08_GB_20251015
#the standard database with archaea, bacteria, viral, plasmid and human sequences - capped at 8 gigabytes
#(Wood, Lu and Langmead, 2019)
```

## Data acquisition and quality control assessment
```bash
mkdir -p ~/1-raw
cd ~/1-raw
wget -r -np -nH --cut-dirs=3 -A "*.fastq.gz" https://cgr.liv.ac.uk/454/acdarby/LIFE750/
#This downloads the three sets of data used in this study: K1_R1, K1_R2 , K2_R1, K2_R2, W1_R1 and W2_R2 (all fastq.gz), where K1 and K2 are two different individuals on a Korean diet and W1 is an individual on a Western diet. R1 signifies read-pair1 and E2 read-pair 2 as these were illumina read-pairs
mkdir R1_fastqc
mkdir R2_fastqc
fastqc -t 4 -o R1_fastqc *R1.fastq.gz
fastqc -t 4 -o R2_fastqc *R2.fastq.gz
mkdir R1_fastqc/multiqc
mkdir R2_fastqc/multiqc
multiqc -o R1_fastqc/multiqc R1_fastqc
multiqc -o R2_fastqc/multiqc R2_fastqc
#(Ewels et al., 2016),(s-andrews, 2018)
```

## Visualisation of read pairs 1 and 2 initial quality 
```bash
firefox 1-raw/R1_fastqc/multiqc/multiqc_report.html \
1-raw/R2_fastqc/multiqc/multiqc_report.html & #to run in background
```

## Trimming read pairs 
```bash
cd ~ #return to home directory
mkdir 2-Trimmed
cd 2-Trimmed
trim_galore --paired --quality 20 --stringency 4 \
~/1-raw/K1_R1.fastq.gz ~/1-raw/K1_R2.fastq.gz
trim_galore --paired --quality 20 --stringency 4 \
~/1-raw/K2_R1.fastq.gz ~/1-raw/K2_R2.fastq.gz
trim_galore --paired --quality 20 --stringency 4 \
~/1-raw/W1_R1.fastq.gz ~/1-raw/W1_R2.fastq.gz
#all sequences generated were re-named in the following manner
mv K1_R1_val_1.fq.gz K1_R1.fq.gz
mkdir R1_fastqc
mkdir R2_fastqc
fastqc -t 4 -o R1_fastqc *R1.fq.gz
fastqc -t 4 -o R2_fastqc *R2.fq.gz
mkdir R1_fastqc/multiqc
mkdir R2_fastqc/multiqc
multiqc -o R1_fastqc/multiqc R1_fastqc
multiqc -o R2_fastqc/multiqc R2_fastqc
#(Ewels et al., 2016),(s-andrews, 2018),(Krueger, 2024)
```

## Quality control read pairs 1  and 2 post-trimming
```bash
firefox 2-Trimmed/R1_fastqc/multiqc/multiqc_report.html \
2-Trimmed/R2_fastqc/multiqc/multiqc_report.html &
```

## Kraken2
```bash
cd ~
mkdir 3-Taxonomy
cd 3-Taxonomy
kraken2 --paired \
--db ~/kraken2_db/k2_standard_08_GB_20251015 \
--output K1.kraken \
--report K1.kreport2 \
~/2-Trimmed/K1_R1.fq.gz ~/2-Trimmed/K1_R2.fq.gz

kraken2 --paired \
--db ~/kraken2_db/k2_standard_08_GB_20251015 \
--output K1.kraken \
--report K1.kreport2 \
~/2-Trimmed/K2_R1.fq.gz ~/2-Trimmed/K2_R2.fq.gz

kraken2 --paired \
--db ~/kraken2_db/k2_standard_08_GB_20251015 \
--output K1.kraken \
--report K1.kreport2 \
~/2-Trimmed/W1_R1.fq.gz ~/2-Trimmed/W1_R2.fq.gz

ktImportTaxonomy -o kraken2.krona.html *.kreport2

firefox kraken2.krona.html
#(Ondov, Bergman and Phillippy, 2011),(Wood, Lu and Langmead, 2019)
```

## Bracken 
```bash
#kraken2 kreport files are used to run Bracken
bracken -d ~/kraken2_db/k2_standard_08_GB_20251015 \
  -i K1.kreport2 \
  -o K1.bracken \
  -w K1.breport2 \
  -r 100 \
  -l S \
  -t 5
  
bracken -d ~/kraken2_db/k2_standard_08_GB_20251015 \
  -i K2.kreport2 \
  -o K2.bracken \
  -w K2.breport2 \
  -r 100 \
  -l S \
  -t 5

bracken -d ~/kraken2_db/k2_standard_08_GB_20251015 \
  -i K2.kreport2 \
  -o K2.bracken \
  -w K2.breport2 \
  -r 100 \
  -l S \
  -t 5

combine_bracken_outputs.py --files K1.bracken K2.bracken W1.bracken --o all.bracken.three_indvs

seq -s , 4 2 50
bracken_num_columns=$(seq -s , 4 2 50)
echo $bracken_num_columns

cut -f 1,$bracken_num_columns all.bracken.three_indvs > all_num_threeindvs.bracken

nano all_num_threeindvs.bracken
#diet, 2 K's and a W were written as the files header, with tab-separation. The file was renamed all_num_threeindvs.lefse.bracken

lefse_format_input.py all_num_threeindvs.lefse.bracken all_num_threeindvs.lefse -c 1 -u 2 -o 1000000

lefse_run.py all_num_threeindvs.lefse all_num_threeindvs.lefse.out
#Number of significantly discriminative features: 0 ( 0 ) before internal wilcoxon
#No features with significant differences between the two classes
#Number of discriminative features with abs LDA score > 2.0 : 0

lefse_plot_res.py --dpi 200 --format png all_num_threeindvs.lefse.out threeindvs.png
#No differentially abundant features found in all_num_threeindvs.lefse.out
#No png output 
#(Segata et al., 2011),(Lu et al., 2017)
```

## Bracken with additional files
```bash
Lefse file created as series of combined bracken outputs from 6 Korean individuals (with 2 datapoints each, before and after diet) and 6 Western individuals (before and after diet datapoints) was downloaded. Data originated from this study: https://www.ebi.ac.uk/metagenomics/studies/MGYS00000394#overview

wget hhtps://cgr.liv.ac.uk/454/acdarby/LIFE750/lefse/all_num.lefse.bracken
#lefse file created from first using above workflow to combine the 12 K and 12 W bracken files, by Professor Darby
mv all_num.lefse.bracken 3-Taxonomy

lefse_format_input.py all_num.lefse.bracken all_num.lefse -c 1 -u 2 -o 1000000
lefse_run.py all_num.lefse all_num.lefse.out
lefse_plot_res.py --dpi 200 --format png all_num.lefse.out biomarkers.png
firefox biomarkers.png
#(Segata et al., 2011)
```

## References Part 1
```
#Where a paper citation is not provided, a link to the relevant github in citation format is.
Ewels, P., Magnusson, M., Lundin, S. and Käller, M. (2016). MultiQC: summarize analysis results for multiple tools and samples in a single report. Bioinformatics, 32(19), pp.3047–3048. doi:https://doi.org/10.1093/bioinformatics/btw354.
Krueger, F. (2024). FelixKrueger/TrimGalore. [online] GitHub. Available at: https://github.com/FelixKrueger/TrimGalore?tab=readme-ov-file.
Lu, J., Breitwieser, F.P., Thielen, P. and Salzberg, S.L. (2017). Bracken: estimating species abundance in metagenomics data. PeerJ Computer Science, 3, p.e104. doi:https://doi.org/10.7717/peerj-cs.104.
Ondov, B.D., Bergman, N.H. and Phillippy, A.M. (2011). Interactive metagenomic visualization in a Web browser. BMC Bioinformatics, 12(1). doi:https://doi.org/10.1186/1471-2105-12-385.
s-andrews (2018). s-andrews/FastQC. [online] GitHub. Available at: https://github.com/s-andrews/FastQC.
Segata, N., Izard, J., Waldron, L., Gevers, D., Miropolsky, L., Garrett, W.S. and Huttenhower, C. (2011). Metagenomic biomarker discovery and explanation. Genome Biol, [online] 12(6), p.R60. doi:https://doi.org/10.1186/gb-2011-12-6-r60.
Wood, D.E., Lu, J. and Langmead, B. (2019). Improved metagenomic analysis with Kraken 2. Genome Biology, 20(1). doi:https://doi.org/10.1186/s13059-019-1891-0.
```

## Part 2
## Stitching read pairs
```bash
cd ~
mkdir 5-Stitched
cd ~/5-Stitched
flash -o K1 -z -t 12 -d . \
~/2-Trimmed/K1_R1.fq.gz ~/2-Trimmed/K1_R2.fq.gz
#set threads (-t) to be suitable to used device, mine was fine with 12 threads as has 16 cores.
#-d = output to current directory (. means current directory)
#(Magoc and Salzberg, 2011)
```

## Megahit 
```bash
cd ~
mkdir 6-Assembly
cd ~/6-Assembly
megahit \
-r ~/5-Stitched/K1.extendedFrags.fastq.gz \
-1 ~/5-Stitched/K1.notCombined_1.fastq.gz \
-2 ~/5-Stitched/K1.notCombined_2.fastq.gz \
-o K1 \
-t 12 \
--k-list 29,49,69,89,109,129,149,169,189
less K1/final.contigs.fa #review output
#(Li et al., 2015)
```

## Quast
```bash 
#in 6-Assembly folder
mkdir -p quast/K1
quast -o quast/K1 K1/final.contigs.fa
firefox quast/K1/report.html
#(Mikheenko et al., 2018)
```

## Metabat Genome binning
```bash
cd ~
mkdir -p ~/7-Binning/K1
cd ~/7-Binning/K1
bwa index ~/6-Assembly/K1/final.contigs.fa
bwa mem ~/6-Assembly/K1/final.contigs.fa \
~/2-Trimmed/K1_R1.fq.gz ~/2-Trimmed/K1_R1.fq.gz > \
K1.sam #generated in K1 folder
samtools view -bu K1.sam > K1.bam
samtools sort K1.bam > K1.sort.bam
jgi_summarize_bam_contig_depths --outputDepth K1.depth.txt K1.sort.bam
less K1.depth.txt #view output 
mkdir bins (inside 7-Binning/K1)
metabat2 \
--inFile ~/6-Assembly/K1/final.contigs.fa \
--outFIle bins/K1 \
--abdFile K1.depth.txt \
--minContig 1500
#generated 2 bins, K1.1.fa and K1.2.fa
#(Li, 2013),(Kang et al., 2019),(Danecek et al., 2021)
```

# Bakta 
```bash
mamba deactivate
mamba activate shotgun_meta
conda install -c conda-forge -c bioconda bakta
wget https://zenodo.org/records/14916843/files/db-light.tar.xz?download=1
tar -xvsf 'db-light.tar.xz?download=1'

#data was downloaded for this for all individuals, not just K1, K2 and W2 with the following:
wget -r -np -nH --cut-dirs=4 \ https://cgr.liv.ac.uk/454/acdarby/LIFE750/Workshops/Workshop01/8-Annotation/
#downloaded to 7-Binning/K1_fullset/bins, K1.*.fa - * being 1-30 bins 
cd ~
mkdir 7-Binning_cont
cd ~/7-Binning_cont
ls -1 ~/7-Binning/K1_fullset/bins/*fa | while read f ; \
do s=$(basename $f | sed "s/.fa//") ; echo $s ; \
bakta --db ~/bakta_db/db-light\
-o ${s} \
$f
done
#bakta outdated for databases downloaded above, used bakta_db download --output ~ --type light
#with path of ~/bakta_db/db-light

#could use code - 
for f in ~/7-Binning/K1_fullset/bins/*.fa; do
    s=$(basename "$f" .fa)
    echo "$s"
    bakta --db ~/bakta_db/db-light -o "$s" "$f"
done
#error message = AttributeError: module 'pyrodigal' has no attribute 'OrfFinder',
installed pyrodigal version incompatible with bakta
#conda create -n bakta -c conda-forge -c bioconda bakta
#conda activate bakta

ls -1 ~/7-Binning/K1_fullset/bins/*fa | while read f ; \
do s=$(basename $f | sed "s/.fa//") ; echo $s ; \
bakta --db ~/db-light \
-o ${s} \
$f
done
#ran with newer database first downloaded, armfinder error, so ran - amrfinder_update --force_update \
--database /home/james/db-light/amrfinderplus-db
#ran command again 
#This creates 30 gff3 files, which can be parsed for gene names- K1.1 and K1.2 will match the 
#bins which were made by the three individuals K1, K2 and W1 used throughout. 
#(Schwengers et al., 2021)
```

## Part 2 references 
```
Danecek, P., Bonfield, J.K., Liddle, J., Marshall, J., Ohan, V., Pollard, M.O., Whitwham, A., Keane, T., McCarthy, S.A., Davies, R.M. and Li, H. (2021). Twelve years of SAMtools and BCFtools. GigaScience, 10(2). doi:https://doi.org/10.1093/gigascience/giab008.
Kang, D.D., Li, F., Kirton, E., Thomas, A., Egan, R., An, H. and Wang, Z. (2019). MetaBAT 2: an adaptive binning algorithm for robust and efficient genome reconstruction from metagenome assemblies. PeerJ, [online] 7, p.e7359. doi:https://doi.org/10.7717/peerj.7359.
Li, D., Liu, C.-M., Luo, R., Sadakane, K. and Lam, T.-W. (2015). MEGAHIT: an ultra-fast single-node solution for large and complex metagenomics assembly via succinct de Bruijn graph. Bioinformatics, [online] 31(10), pp.1674–1676. doi:https://doi.org/10.1093/bioinformatics/btv033.
Li, H. (2013). Aligning sequence reads, clone sequences and assembly contigs with BWA-MEM. [online] arXiv.org. Available at: https://arxiv.org/abs/1303.3997.
Magoc, T. and Salzberg, S.L. (2011). FLASH: fast length adjustment of short reads to improve genome assemblies. Bioinformatics, 27(21), pp.2957–2963. doi:https://doi.org/10.1093/bioinformatics/btr507.
Mikheenko, A., Prjibelski, A., Saveliev, V., Antipov, D. and Gurevich, A. (2018). Versatile genome assembly evaluation with QUAST-LG. Bioinformatics, 34(13), pp.i142–i150. doi:https://doi.org/10.1093/bioinformatics/bty266.
Schwengers, O., Jelonek, L., Dieckmann, M.A., Beyvers, S., Blom, J. and Goesmann, A. (2021). Bakta: rapid and standardized annotation of bacterial genomes via alignment-free sequence identification. Microbial Genomics, 7(11). doi:https://doi.org/10.1099/mgen.0.000685.
```
